wiki/python/ESTIMATOR_GUIDE.md

```diff
@@ -0,0 +1,833 @@
+```python
+def my_func(self, X):
+   with cuml.using_ouput_type("mirror"):
+      with cupy.cuda.cupy_using_allocator(rmm.rmm_cupy_allocator):
+         # Set the input properties
+         self._set_base_attributes(output_type=X, n_features=X)
+
+         # Do actual calculation returning an array-like object
+         ret_val = self._my_func(X)
+
+         # Get the output type
+         output_type = self._get_output_type(X)
+
+         # Convert array-like to CumlArray
+         ret_val = input_to_cuml_array(ret_val, order="K").array
+
+         # Convert CumlArray to desired output_type
+         return ret_val.to_output(output_type)
+```
+
+Keep the above pseudocode in mind when working with these decorators since their goal is to replace many of these repetitive functions.
+
+### Decorator Defaults
+
+Every function in `cuml` is slightly different and some `fit()` functions may need to set the `target_dtype` or some `predict()` functions may need to skip getting the output type. To handle these situations, all of the decorators take arguments to configure their functionality.
+
+Since the decorator's functionality is very similar, so are their arguments. All of the decorators take similar arguments that will be outlined below.
+
+| Argument | Type | Default | Meaning | 
+| :-----------: | :-----------: | :-----------: | :----------- |
+| `input_arg` | `str` | `'X'` or 1st non-self argument | Determines which input argument to use for `_set_output_type()` and `_set_n_features_in()` |
+| `target_arg` | `str` | `'y'` or 2nd non-self argument | Determines which input argument to use for `_set_target_dtype()` |
+| `set_output_type` | `bool` | Varies | Whether to call `_set_output_type(input_arg)` |
+| `set_output_dtype` | `bool` | `False` | Whether to call `_set_target_dtype(target_arg)` |
+| `set_n_features_in` | `bool` | Varies | Whether to call `_set_n_features_in(input_arg)` |
+| `get_output_type` | `bool` | Varies | Whether to call `_get_output_type(input_arg)` |
+| `get_output_dtype` | `bool` | `False` | Whether to call `_get_target_dtype()` |
+
+An example of how these arguments can be used is below:
+
+**Before:**
+```python
+@with_cupy_rmm
+def predict(self, X, y):
+   # Determine the output type and dtype
+   out_type = self._get_output_type(y)
+   out_dtype = self._get_target_dtype()
+
+   # Convert to CumlArray
+   X_m = input_to_cuml_array(X, order="K").array
+
+   # Call a cuda function
+   someCudaFunction(X_m.ptr)
+
+   # Convert the CudaArray to the desired output
+   return X_m.to_output(output_type=out_type, output_dtype=out_dtype)
+```
+
+**After:**
+```python
+@cuml.internals.api_base_return_array(input_arg="y", get_output_dtype=True)
+def predict(self, X):
+   # Convert to CumlArray
+   X_m = input_to_cuml_array(X, order="K").array
+
+   # Call a cuda function
+   someCudaFunction(X_m.ptr)
+
+   # Convert the CudaArray to the desired output
+   return X_m
+```
+
+#### Before `0.17` and After Comparison
```

Comment:
Love how you moved the section, thanks! It makes is super clear for existing devs to see pre-0.17/post0.17 behavior, love it.

Reply:
Great. thanks.